File: frontend/src/functions/http-client/api/setting.ts
import { UsefulColors } from "@/constants/ui"
import { HttpInstance, Response } from "../instance"
import { AlreadyExists, CascadeResourceExists, NotFound, InvalidRuleIndexError, ResourceNotExist, BuiltinNotWritableError } from "../exceptions"
import { AuthorType } from "./author"
import { TopicType } from "./topic"
import { TaskConfig } from "./find-similar"
import { MetaType } from "./all"

export function createSettingEndpoint(http: HttpInstance): SettingEndpoint {
    return {
        server: {
            get: http.createRequest("/api/setting/server"),
            update: http.createDataRequest("/api/setting/server", "PATCH")
        },
        storage: {
            get: http.createRequest("/api/setting/storage"),
            update: http.createDataRequest("/api/setting/storage", "PATCH")
        },
        meta: {
            get: http.createRequest("/api/setting/meta"),
            update: http.createDataRequest("/api/setting/meta", "PATCH")
        },
        query: {
            get: http.createRequest("/api/setting/query"),
            update: http.createDataRequest("/api/setting/query", "PATCH")
        },
        import: {
            get: http.createRequest("/api/setting/import"),
            update: http.createDataRequest("/api/setting/import", "PATCH")
        },
        source: {
            site: {
                list: http.createRequest("/api/setting/source/sites"),
                listBuiltins: http.createRequest("/api/setting/source/sites/builtins"),
                create: http.createDataRequest("/api/setting/source/sites", "POST"),
                get: http.createPathRequest(name => `/api/setting/source/sites/${encodeURIComponent(name)}`),
                update: http.createPathDataRequest(name => `/api/setting/source/sites/${encodeURIComponent(name)}`, "PUT"),
                delete: http.createPathRequest(name => `/api/setting/source/sites/${encodeURIComponent(name)}`, "DELETE"),
            }
        },
        findSimilar: {
            get: http.createRequest("/api/setting/find-similar"),
            update: http.createDataRequest("/api/setting/find-similar", "PATCH")
        }
    }
}

export interface SettingEndpoint {
    /**
     * 设置：后台服务本身相关的选项。
     */
    server: {
        /**
         * 查看。
         */
        get(): Promise<Response<ServerOption>>
        /**
         * 更改。
         */
        update(form: ServerOptionUpdateForm): Promise<Response<unknown>>
    }
    /**
     * 设置：存储相关选项。
     */
    storage: {
        /**
         * 查看。
         */
        get(): Promise<Response<StorageOption>>
        /**
         * 更改。
         */
        update(form: StorageOptionUpdateForm): Promise<Response<unknown>>
    }
    /**
     * 设置：基本元数据相关的选项。
     */
    meta: {
        /**
         * 查看。
         */
        get(): Promise<Response<MetaOption>>
        /**
         * 更改。
         */
        update(form: MetaOptionUpdateForm): Promise<Response<unknown>>
    }
    /**
     * 设置：查询相关的选项。
     */
    query: {
        /**
         * 查看。
         */
        get(): Promise<Response<QueryOption>>
        /**
         * 更改。
         */
        update(form: QueryOptionUpdateForm): Promise<Response<unknown>>
    }
    /**
     * 设置：导入相关选项。
     */
    import: {
        /**
         * 查看。
         */
        get(): Promise<Response<ImportOption>>
        /**
         * 更改。
         * @exception NOT_EXIST("site", siteName) rules[].site不在sites列表中存在时报告此错误。
         * @exception INVALID_RULE_INDEX rules[]中part/partName/extras的group配置与site定义不匹配时报告此错误。
         */
        update(form: ImportOptionUpdateForm): Promise<Response<unknown, ResourceNotExist<"site", string> | InvalidRuleIndexError>>
    }
    /**
     * 设置：来源数据相关的选项。
     */
    source: {
        /**
         * 来源网站列表。
         */
        site: {
            /**
             * 查看列表。
             */
            list(): Promise<Response<Site[]>>
            /**
             * 查看内建列表。
             */
            listBuiltins(): Promise<Response<Site[]>>
            /**
             * 新增一个来源网站。
             * @exception ALREADY_EXISTS
             */
            create(form: SiteCreateForm): Promise<Response<null, AlreadyExists<"site", "name", string>>>
            /**
             * 查看单个项。
             * @exception NOT_FOUND 此项不存在。
             */
            get(name: string): Promise<Response<Site, NotFound>>
            /**
             * 更改项。
             * @exception NOT_FOUND 此项不存在。
             */
            update(name: string, form: SiteUpdateForm): Promise<Response<null, NotFound | BuiltinNotWritableError>>
            /**
             * 删除项。
             * @exception NOT_FOUND 此项不存在。
             * @exception CASCADE_RESOURCE_EXISTS("Illust"|"ImportImage"|"TrashedImage"|"SourceAnalyseRule", "site", string) 存在级联资源，无法删除。
             */
            delete(name: string): Promise<Response<null, NotFound | CascadeResourceExists<"Illust" | "ImportImage" | "TrashedImage" | "SourceAnalyseRule", "site", string>>>
        }
    }
    /**
     * 设置：相关项查找相关选项。
     */
    findSimilar: {
        /**
         * 查看。
         */
        get(): Promise<Response<FindSimilarOption>>
        /**
         * 更改。
         */
        update(form: FindSimilarOptionUpdateForm): Promise<Response<unknown>>
    }
    
}

export interface ServerOption {
    /**
     * 后台服务建议使用的端口。
     * null表示没有建议，由它自己选择端口。
     * 使用整数+逗号(,)+横线(-)表示建议的范围。
     * 这个参数没有强制检查，如果写错，则在检测时不生效。
     */
    port: string | null
    /**
     * 后台服务额外的固定Token。
     * 此Token与生成Token并行可用。使用此Token可在其他位置访问后台服务。
     */
    token: string | null
    /**
     * 在各处有关日期的判定中，每天的日期范围的推迟时间量。
     */
    timeOffsetHour: number | null
}

export type ServerOptionUpdateForm = Partial<ServerOption>

export interface StorageOption {
    /**
     * 文件存储使用的存储路径。
     * null表示使用基于userData的默认存储路径，并自行管理；其他值表示使用自定义路径，并需要自行确保路径可用。
     */
    storagePath: string | null
    /**
     * 自动清理已删除的项。
     */
    autoCleanTrashes: boolean
    /**
     * 自动清理已删除项的间隔天数。
     */
    autoCleanTrashesIntervalDay: number
    /**
     * 区块最大可储存的容量。
     */
    blockMaxSizeMB: number
    /**
     * 区块最大可储存的数量。
     */
    blockMaxCount: number
}

export type StorageOptionUpdateForm = Partial<StorageOption>

export interface MetaOption {
    /**
     * 对相关元数据做更改后自动清除对应的tagme标记。
     */
    autoCleanTagme: boolean
    /**
     * 以角色标签为主题核心。
     */
    onlyCharacterTopic: boolean
    /**
     * 根据推导得到的父标签解决子标签冲突。用于解决character多义映射的情况。这个选项会在“导入自动映射”和“根据来源标签批量设置标签”功能中生效。
     */
    resolveTagConflictByParent: boolean
    /**
     * 当创建新集合或向集合添加新项时，允许指定分区，将不在此分区的项聚集到此分区中。
     */
    centralizeCollection: boolean
    /**
     * 自动微调排序时间。
     */
    tuningOrderTime: boolean
    /**
     * 对orderTime的变更将会自动同步至partitionTime。
     */
    bindingPartitionWithOrderTime: boolean
    /**
     * topic根据type区分的颜色。
     */
    topicColors: {[key in TopicType]: UsefulColors}
    /**
     * author根据type区分的颜色。
     */
    authorColors: {[key in AuthorType]: UsefulColors}
}

export type MetaOptionUpdateForm = Partial<MetaOption>

export interface QueryOption {
    /**
     * 识别中文字符。
     */
    chineseSymbolReflect: boolean
    /**
     * 将有限字符串中的下划线转义为空格。
     */
    translateUnderscoreToSpace: boolean
    /**
     * 查询元数据库中的标签项时，每一次查询的数量上限。
     * @min 1
     */
    queryLimitOfQueryItems: number
    /**
     * 每一个项中的元素总数的警告阈值。
     * @min 2
     */
    warningLimitOfUnionItems: number
    /**
     * 总项数的警告阈值。
     * @min 2
     */
    warningLimitOfIntersectItems: number
}

export type QueryOptionUpdateForm = Partial<QueryOption>

export interface ImportOption {
    /**
     * 导入时，自动从文件分析source信息。
     */
    autoAnalyseSourceData: boolean
    /**
     * 阻止没有来源信息的项被导入。这有利于保证所有导入项都有来源。
     */
    preventNoneSourceData: boolean
    /**
     * 在文件导入时，根据已设置的来源和映射规则，自动映射并添加元数据标签。
     */
    autoReflectMetaTag: boolean
    /**
     * 启用哪些元数据标签类型的映射。
     */
    reflectMetaTagType: MetaType[]
    /**
     * 在文件导入时，自动对那些特定格式的、容量较大的图像进行格式转换，以在不损失质量的前提下减少其大小。
     */
    autoConvertFormat: boolean
    /**
     * PNG类型的自动转换阈值大小
     */
    autoConvertPNGThresholdSizeMB: number
    /**
     * 导入时，自动设定meta tag的tagme。
     */
    setTagmeOfTag: boolean
    /**
     * 导入时，使用哪种属性设置create time。
     */
    setOrderTimeBy: OrderTimeType
    /**
     * source分析的规则列表。
     */
    sourceAnalyseRules: SourceAnalyseRule[]
}

export type ImportOptionUpdateForm = Partial<ImportOption>

export interface FindSimilarOption {
    autoFindSimilar: boolean
    autoTaskConf: TaskConfig | null
    defaultTaskConf: TaskConfig
}

export type FindSimilarOptionUpdateForm = Partial<FindSimilarOption>

/**
 * 一条source解析规则。
 */
export interface SourceAnalyseRule {
    site: string
    regex: string
    idGroup: string
    partGroup: string | null
    partNameGroup: string | null
    extras: SourceAnalyseRuleExtra[] | null
}

export interface SourceAnalyseRuleExtra {
    group: string
    target: SourceAnalyseRuleExtraTarget
    optional: boolean
    tagType: string | null
    additionalInfoField: string | null
    translateUnderscoreToSpace: boolean
}

export interface Site {
    /**
     * 网站的识别名称。
     */
    name: string
    /**
     * 网站的显示名称。
     */
    title: string
    /**
     * 是否是内建网站。
     */
    isBuiltin: boolean
    /**
     * 此网站的ID模式。
     */
    idMode: SiteIdMode
    /**
     * 此网站是否拥有分页。
     * @default NO
     */
    partMode: SitePartMode
    /**
     * 此网站可接受的元数据条目。
     */
    additionalInfo: {field: string, label: string}[]
    /**
     * 根据元数据id与附加信息，自动生成links的规则列表。
     */
    sourceLinkRules: string[]
    /**
     * 此站点可用的标签类型。
     */
    tagTypes: string[]
    /**
     * 此站点的标签类型向元数据类型的映射。
     */
    tagTypeMappings: Record<string, Exclude<TopicType, "UNKNOWN"> | Exclude<AuthorType, "UNKNOWN"> | "TAG">
}

export interface SiteCreateForm extends SiteUpdateForm {
    name: string
    idMode?: SiteIdMode
    partMode?: SitePartMode
}

export interface SiteUpdateForm {
    /**
     * 在列表中的排序顺序，从0开始。
     * @default 追加到末尾
     */
    ordinal?: number
    title?: string | null
    additionalInfo?: {field: string, label: string}[]
    sourceLinkRules?: string[]
    tagTypes?: string[]
    tagTypeMappings?: Site["tagTypeMappings"]
}

export type OrderTimeType = "IMPORT_TIME" | "CREATE_TIME" | "UPDATE_TIME"

export type SourceAnalyseRuleExtraTarget = "TITLE" | "DESCRIPTION" | "ADDITIONAL_INFO" | "TAG" | "BOOK" | "RELATION"

export type SiteIdMode = "NUMBER" | "STRING"

export type SitePartMode = "NO" | "PAGE" | "PAGE_WITH_NAME"

File: cli/src/module/api/setting.rs
use std::{error::Error, collections::HashMap};
use reqwest::Method;
use serde::{Deserialize, Serialize};
use crate::module::server::ServerManager;


pub struct SettingModule<'t> {
    server_manager: &'t ServerManager
}

impl <'t> SettingModule<'t> {
    pub fn new(server_manager: &'t ServerManager) -> SettingModule {
        SettingModule { server_manager }        
    }
    pub async fn _get_server_option(&mut self) -> Result<ServerOption, Box<dyn Error>> {
        self.server_manager.req(Method::GET, "/api/setting/server").await
    }
    pub async fn _get_storage_option(&mut self) -> Result<StorageOption, Box<dyn Error>> {
        self.server_manager.req(Method::GET, "/api/setting/file").await
    }
    pub async fn _get_find_similar_option(&mut self) -> Result<FindSimilarOption, Box<dyn Error>> {
        self.server_manager.req(Method::GET, "/api/setting/find-similar").await
    }
    pub async fn _get_query_option(&mut self) -> Result<QueryOption, Box<dyn Error>> {
        self.server_manager.req(Method::GET, "/api/setting/query").await
    }
    pub async fn _get_meta_option(&mut self) -> Result<MetaOption, Box<dyn Error>> {
        self.server_manager.req(Method::GET, "/api/setting/meta").await
    }
    pub async fn _get_import_option(&mut self) -> Result<ImportOption, Box<dyn Error>> {
        self.server_manager.req(Method::GET, "/api/setting/import").await
    }
    pub async fn _get_source_sites(&mut self) -> Result<Vec<SourceSite>, Box<dyn Error>> {
        self.server_manager.req(Method::GET, "/api/setting/source/sites").await
    }
    pub async fn set_server_option(&mut self, bulks: &ServerOptionUpdateForm) -> Result<(), Box<dyn Error>> {
        let body = serde_json::to_value(bulks)?;
        self.server_manager.req_without_res(Method::PATCH, "/api/setting/server", body).await
    }
    pub async fn set_storage_option(&mut self, bulks: &StorageOptionUpdateForm) -> Result<(), Box<dyn Error>> {
        let body = serde_json::to_value(bulks)?;
        self.server_manager.req_without_res(Method::PATCH, "/api/setting/file", body).await
    }
    pub async fn set_find_similar_option(&mut self, bulks: &FindSimilarOptionUpdateForm) -> Result<(), Box<dyn Error>> {
        let body = serde_json::to_value(bulks)?;
        self.server_manager.req_without_res(Method::PATCH, "/api/setting/find-similar", body).await
    }
    pub async fn set_query_option(&mut self, bulks: &QueryOptionUpdateForm) -> Result<(), Box<dyn Error>> {
        let body = serde_json::to_value(bulks)?;
        self.server_manager.req_without_res(Method::PATCH, "/api/setting/query", body).await
    }
    pub async fn set_meta_option(&mut self, bulks: &MetaOptionUpdateForm) -> Result<(), Box<dyn Error>> {
        let body = serde_json::to_value(bulks)?;
        self.server_manager.req_without_res(Method::PATCH, "/api/setting/meta", body).await
    }
    pub async fn set_import_option(&mut self, bulks: &ImportOptionUpdateForm) -> Result<(), Box<dyn Error>> {
        let body = serde_json::to_value(bulks)?;
        self.server_manager.req_without_res(Method::PATCH, "/api/setting/import", body).await
    }
    pub async fn set_source_sites(&mut self, bulks: &Vec<SourceSiteUpdateForm>) -> Result<(), Box<dyn Error>> {
        let body = serde_json::to_value(bulks)?;
        self.server_manager.req_without_res(Method::PUT, "/api/setting/source/sites", body).await
    }
}

#[derive(Deserialize)]
pub struct ServerOption {
    #[serde(rename = "port")]
    pub port: Option<String>,
    #[serde(rename = "token")]
    pub token: Option<String>,
    #[serde(rename = "timeOffsetHour")]
    pub time_offset_hour: Option<i32>
}

#[derive(Deserialize)]
pub struct StorageOption {
    #[serde(rename = "storagePath")]
    pub storage_path: Option<String>,
    #[serde(rename = "autoCleanTrashes")]
    pub auto_clean_trashes: bool,
    #[serde(rename = "autoCleanTrashesIntervalDay")]
    pub auto_clean_trashes_interval_day: i32,
    #[serde(rename = "blockMaxSizeMB")]
    pub block_max_size: i64,
    #[serde(rename = "blockMaxCount")]
    pub block_max_count: i32
}

#[derive(Deserialize)]
pub struct FindSimilarOption {
    #[serde(rename = "autoFindSimilar")]
    pub auto_find_similar: bool,
    #[serde(rename = "autoTaskConf")]
    pub auto_task_conf: Option<FindSimilarTaskConfig>,
    #[serde(rename = "defaultTaskConf")]
    pub default_task_conf: FindSimilarTaskConfig,
}

#[derive(Deserialize)]
pub struct QueryOption {
    #[serde(rename = "chineseSymbolReflect")]
    pub chinese_symbol_reflect: bool,
    #[serde(rename = "translateUnderscoreToSpace")]
    pub translate_underscore_to_space: bool,
    #[serde(rename = "queryLimitOfQueryItems")]
    pub query_limit_of_query_items: i32,
    #[serde(rename = "warningLimitOfUnionItems")]
    pub warning_limit_of_union_items: i32,
    #[serde(rename = "warningLimitOfIntersectItems")]
    pub warning_limit_of_intersect_items: i32
}

#[derive(Deserialize)]
pub struct MetaOption {
    #[serde(rename = "autoCleanTagme")]
    pub auto_clean_tagme: bool,
    #[serde(rename = "onlyCharacterTopic")]
    pub only_character_topic: bool,
    #[serde(rename = "resolveTagConflictByParent")]
    pub resolve_tag_conflict_by_parent: bool,
    #[serde(rename = "bindingPartitionWithOrderTime")]
    pub binding_partition_with_order_time: bool,
    #[serde(rename = "tuningOrderTime")]
    pub tuning_order_time: bool,
    #[serde(rename = "centralizeCollection")]
    pub centralize_collection: bool,
    #[serde(rename = "topicColors")]
    pub topic_colors: HashMap<String, String>,
    #[serde(rename = "authorColors")]
    pub author_colors: HashMap<String, String>
}

#[derive(Deserialize)]
pub struct ImportOption {
    #[serde(rename = "autoAnalyseSourceData")]
    pub auto_analyse_source_data: bool,
    #[serde(rename = "preventNoneSourceData")]
    pub prevent_none_source_data: bool,
    #[serde(rename = "autoReflectMetaTag")]
    pub auto_reflect_meta_tag: bool,
    #[serde(rename = "reflectMetaTagType")]
    pub reflect_meta_tag_type: Vec<String>,
    #[serde(rename = "autoConvertFormat")]
    pub auto_convert_format: bool,
    #[serde(rename = "autoConvertPNGThresholdSizeMB")]
    pub auto_convert_png_threshold_size_mb: i64,
    #[serde(rename = "setTagmeOfTag")]
    pub set_tagme_of_tag: bool,
    #[serde(rename = "setTagmeOfSource")]
    pub set_tagme_of_source: bool,
    #[serde(rename = "setOrderTimeBy")]
    pub set_order_time_by: String,
    #[serde(rename = "sourceAnalyseRules")]
    pub source_analyse_rules: Vec<SourceAnalyseRule>
}

#[derive(Deserialize)]
pub struct SourceSite {
    pub name: String,
    pub title: String,
    #[serde(rename = "idMode")]
    pub id_mode: String,
    #[serde(rename = "partMode")]
    pub part_mode: String,
    #[serde(rename = "additionalInfo")]
    pub additional_info: Vec<AdditionalInfo>,
    #[serde(rename = "sourceLinkRules")]
    pub source_link_rules: Vec<String>,
    #[serde(rename = "tagTypes")]
    pub tag_types: Vec<String>,
    #[serde(rename = "tagTypeMappings")]
    pub tag_type_mappings: Vec<HashMap<String, String>>
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct ServerOptionUpdateForm {
    #[serde(rename = "port", skip_serializing_if = "Option::is_none")]
    pub port: Option<String>,
    #[serde(rename = "token", skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
    #[serde(rename = "timeOffsetHour", alias = "time_offset_hour", skip_serializing_if = "Option::is_none")]
    pub time_offset_hour: Option<i32>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct StorageOptionUpdateForm {
    #[serde(rename = "storagePath", skip_serializing_if = "Option::is_none")]
    pub storage_path: Option<String>,
    #[serde(rename = "autoCleanTrashes", alias = "auto_clean_trashes", skip_serializing_if = "Option::is_none")]
    pub auto_clean_trashes: Option<bool>,
    #[serde(rename = "autoCleanTrashesIntervalDay", alias = "auto_clean_trashes_interval_day", skip_serializing_if = "Option::is_none")]
    pub auto_clean_trashes_interval_day: Option<i32>,
    #[serde(rename = "blockMaxSizeMB", alias = "block_max_size", skip_serializing_if = "Option::is_none")]
    pub block_max_size: Option<i64>,
    #[serde(rename = "blockMaxCount", alias = "block_max_count", skip_serializing_if = "Option::is_none")]
    pub block_max_count: Option<i32>
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct FindSimilarOptionUpdateForm {
    #[serde(rename = "autoFindSimilar", alias = "auto_find_similar", skip_serializing_if = "Option::is_none")]
    pub auto_find_similar: Option<bool>,
    #[serde(rename = "autoTaskConf", alias = "auto_task_conf", skip_serializing_if = "Option::is_none")]
    pub auto_task_conf: Option<FindSimilarTaskConfig>,
    #[serde(rename = "defaultTaskConf", alias = "default_task_conf", skip_serializing_if = "Option::is_none")]
    pub default_task_conf: Option<FindSimilarTaskConfig>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct QueryOptionUpdateForm {
    #[serde(rename = "chineseSymbolReflect", alias = "chinese_symbol_reflect", skip_serializing_if = "Option::is_none")]
    pub chinese_symbol_reflect: Option<bool>,
    #[serde(rename = "translateUnderscoreToSpace", alias = "translate_underscore_to_space", skip_serializing_if = "Option::is_none")]
    pub translate_underscore_to_space: Option<bool>,
    #[serde(rename = "queryLimitOfQueryItems", alias = "query_limit_of_query_items", skip_serializing_if = "Option::is_none")]
    pub query_limit_of_query_items: Option<i32>,
    #[serde(rename = "warningLimitOfUnionItems", alias = "warning_limit_of_union_items", skip_serializing_if = "Option::is_none")]
    pub warning_limit_of_union_items: Option<i32>,
    #[serde(rename = "warningLimitOfIntersectItems", alias = "warning_limit_of_intersect_items", skip_serializing_if = "Option::is_none")]
    pub warning_limit_of_intersect_items: Option<i32>
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct MetaOptionUpdateForm {
    #[serde(rename = "autoCleanTagme", alias = "auto_clean_tagme", skip_serializing_if = "Option::is_none")]
    pub auto_clean_tagme: Option<bool>,
    #[serde(rename = "onlyCharacterTopic", alias = "only_character_topic", skip_serializing_if = "Option::is_none")]
    pub only_character_topic: Option<bool>,
    #[serde(rename = "resolveTagConflictByParent", alias = "resolve_tag_conflict_by_parent", skip_serializing_if = "Option::is_none")]
    pub resolve_tag_conflict_by_parent: Option<bool>,
    #[serde(rename = "bindingPartitionWithOrderTime", alias = "binding_partition_with_order_time", skip_serializing_if = "Option::is_none")]
    pub binding_partition_with_order_time: Option<bool>,
    #[serde(rename = "tuningOrderTime", alias = "tuning_order_time", skip_serializing_if = "Option::is_none")]
    pub tuning_order_time: Option<bool>,
    #[serde(rename = "centralizeCollection", alias = "centralize_collection", skip_serializing_if = "Option::is_none")]
    pub centralize_collection: Option<bool>,
    #[serde(rename = "topicColors", alias = "topic_colors", skip_serializing_if = "Option::is_none")]
    pub topic_colors: Option<HashMap<String, String>>,
    #[serde(rename = "authorColors", alias = "author_colors", skip_serializing_if = "Option::is_none")]
    pub author_colors: Option<HashMap<String, String>>
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct ImportOptionUpdateForm {
    #[serde(rename = "autoAnalyseSourceData", alias = "auto_analyse_source_data", skip_serializing_if = "Option::is_none")]
    pub auto_analyse_source_data: Option<bool>,
    #[serde(rename = "preventNoneSourceData", alias = "prevent_none_source_data", skip_serializing_if = "Option::is_none")]
    pub prevent_none_source_data: Option<bool>,
    #[serde(rename = "autoReflectMetaTag", alias = "auto_reflect_meta_tag", skip_serializing_if = "Option::is_none")]
    pub auto_reflect_meta_tag: Option<bool>,
    #[serde(rename = "reflectMetaTagType", alias = "reflect_meta_tag_type", skip_serializing_if = "Option::is_none")]
    pub reflect_meta_tag_type: Option<Vec<String>>,
    #[serde(rename = "autoConvertFormat", alias = "auto_convert_format", skip_serializing_if = "Option::is_none")]
    pub auto_convert_format: Option<bool>,
    #[serde(rename = "autoConvertPNGThresholdSizeMB", alias = "auto_convert_png_threshold_size_mb", skip_serializing_if = "Option::is_none")]
    pub auto_convert_png_threshold_size_mb: Option<i64>,
    #[serde(rename = "setTagmeOfTag", alias = "set_tagme_of_tag", skip_serializing_if = "Option::is_none")]
    pub set_tagme_of_tag: Option<bool>,
    #[serde(rename = "setTagmeOfSource", alias = "set_tagme_of_source", skip_serializing_if = "Option::is_none")]
    pub set_tagme_of_source: Option<bool>,
    #[serde(rename = "setOrderTimeBy", alias = "set_order_time_by", skip_serializing_if = "Option::is_none")]
    pub set_order_time_by: Option<String>,
    #[serde(rename = "sourceAnalyseRules", alias = "source_analyse_rules", skip_serializing_if = "Option::is_none")]
    pub source_analyse_rules: Option<Vec<SourceAnalyseRule>>
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct SourceSiteUpdateForm {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(rename = "idMode", alias = "id_mode", skip_serializing_if = "Option::is_none")]
    pub id_mode: Option<String>,
    #[serde(rename = "partMode", alias = "part_mode", skip_serializing_if = "Option::is_none")]
    pub part_mode: Option<String>,
    #[serde(rename = "additionalInfo", alias = "additional_info", skip_serializing_if = "Option::is_none")]
    pub additional_info: Option<Vec<AdditionalInfo>>,
    #[serde(rename = "sourceLinkRules", alias = "source_link_rules", skip_serializing_if = "Option::is_none")]
    pub source_link_rules: Option<Vec<String>>,
    #[serde(rename = "tagTypes", alias = "tag_types", skip_serializing_if = "Option::is_none")]
    pub tag_types: Option<Vec<String>>,
    #[serde(rename = "tagTypeMappings", alias = "tag_type_mappings", skip_serializing_if = "Option::is_none")]
    pub tag_type_mappings: Option<HashMap<String, String>>
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct FindSimilarTaskConfig {
    #[serde(rename = "findBySourceIdentity", alias = "find_by_source_identity")]
    pub find_by_source_identity: bool,
    #[serde(rename = "findBySourceRelation", alias = "find_by_source_relation")]
    pub find_by_source_relation: bool,
    #[serde(rename = "findBySourceMark", alias = "find_by_source_mark")]
    pub find_by_source_mark: bool,
    #[serde(rename = "findBySimilarity", alias = "find_by_similarity")]
    pub find_by_similarity: bool,
    #[serde(rename = "filterByOtherImport", alias = "filter_by_other_import")]
    pub filter_by_other_import: bool,
    #[serde(rename = "filterByPartition", alias = "filter_by_partition")]
    pub filter_by_partition: bool,
    #[serde(rename = "filterByTopic", alias = "filter_by_topic")]
    pub filter_by_topic: bool,
    #[serde(rename = "filterByAuthor", alias = "filter_by_author")]
    pub filter_by_author: bool,
    #[serde(rename = "filterBySourceTagType", alias = "filter_by_source_tag_type")]
    pub filter_by_source_tag_type: Vec<FindSimilarTaskConfigSourceTagType>
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct FindSimilarTaskConfigSourceTagType {
    #[serde(rename = "sourceSite", alias = "source_site")]
    pub source_site: String,
    #[serde(rename = "tagType", alias = "tag_type")]
    pub tag_type: String
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct SourceAnalyseRule {
    pub site: String,
    pub regex: String,
    #[serde(rename = "idGroup", alias = "id_group")]
    pub id_group: String,
    #[serde(rename = "partGroup", alias = "part_group")]
    pub part_group: Option<String>,
    #[serde(rename = "partNameGroup", alias = "part_name_group")]
    pub part_name_group: Option<String>,
    pub extras: Option<Vec<SourceAnalyseRuleExtra>>
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct SourceAnalyseRuleExtra {
    pub group: String,
    pub target: String,
    pub optional: bool,
    #[serde(rename = "tagType", alias = "tag_type")]
    pub tag_type: Option<String>,
    #[serde(rename = "additionalInfoField", alias = "additional_info_field")]
    pub additional_info_field: Option<String>,
    #[serde(rename = "translateUnderscoreToSpace", alias = "translate_underscore_to_space")]
    pub translate_underscore_to_space: Option<bool>
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct AdditionalInfo {
    pub field: String,
    pub label: String
}
